stage #1/module #6. IO, NIO/lesson #1. IO/content.md

```diff
@@ -0,0 +1,127 @@
+
+Java I/O Streams support many kinds of data, including simple bytes, primitive data types, localized characters, and objects. Some streams simply pass on data;
+others manipulate and transform the data in useful ways
+
+- [Byte Streams](https://docs.oracle.com/javase/tutorial/essential/io/bytestreams.html) handle I/O of raw binary data.
+- [Character Streams](https://docs.oracle.com/javase/tutorial/essential/io/charstreams.html) handle I/O of character data, automatically handling translation to and from the local character set.
+- [Buffered Streams](https://docs.oracle.com/javase/tutorial/essential/io/buffers.html) optimize input and output by reducing the number of calls to the native API.
+- [Scanning and Formatting](https://docs.oracle.com/javase/tutorial/essential/io/scanfor.html) allows a program to read and write formatted text.
+- [I/O from the Command Line](https://docs.oracle.com/javase/tutorial/essential/io/cl.html) describes the Standard Streams and the Console object.
+- [Data Streams](https://docs.oracle.com/javase/tutorial/essential/io/datastreams.html) handle binary I/O of primitive data type and String values.
+- [Object Streams](https://docs.oracle.com/javase/tutorial/essential/io/objectstreams.html) handle binary I/O of objects.
+
+No matter how they work internally, all streams present the same simple model to programs that use them: A stream is a sequence of data. A program uses an *input stream* to read data from a source, one item at a time:
+![img_1.png](assets/img_1.png)
+
+A program uses an *output stream* to write data to a destination, one item at time:
+![img.png](assets/img.png)
+
+## Byte Streams
+
+Programs use *byte streams* to perform input and output of 8-bit bytes. All byte stream classes are descended from [InputStream](https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html) and [OutputStream](https://docs.oracle.com/javase/8/docs/api/java/io/OutputStream.html).
+There are many byte stream classes. To demonstrate how byte streams work, we'll focus on the file I/O byte streams, [FileInputStream](https://docs.oracle.com/javase/8/docs/api/java/io/FileInputStream.html) and [FileOutputStream](https://docs.oracle.com/javase/8/docs/api/java/io/FileOutputStream.html) which read and write files. Other kinds of byte streams are used in much the same way; they differ mainly in the way they are constructed.
+
+Following is an example which makes use of these two classes to copy an input file into an output file using byte stream, one byte at a time:
+
+| ![img_2.png](assets/img_2.png) |
+|-------------------------|
+| Byte Stream usage       |
+
+*CopyFile* spends most of its time in a simple loop that reads the input stream and writes the output stream, one byte at a time, as shown in the following figure.
+![img_4.png](assets/img_4.png)
+
+Method “read()” reads one element and return it’s value. When operation read is ended, then returned “-1”. Method “write(c)” writes one element into file.
```

Comment:
Could you wrap part of code with `? (`read()`)

Reply:
Done